actors/builtin/miner/miner_state.go

```diff
@@ -939,6 +941,26 @@ func (st *State) AddLockedFunds(store adt.Store, currEpoch abi.ChainEpoch, vesti
 	return nil
 }
 
+// PenalizeFundsInPriorityOrder first unlocks unvested funds from the vesting table.
+// If the target is not yet hit it deducts funds from the (new) available balance.
+// Returns the amount unlocked from the vesting table and the amount taken from current balance.
+func (st *State) PenalizeFundsInPriorityOrder(store adt.Store, currEpoch abi.ChainEpoch, target, availableBalance abi.TokenAmount) (fromVesting abi.TokenAmount, fromBalance abi.TokenAmount, err error) {
+	fromVesting, err = st.UnlockUnvestedFunds(store, currEpoch, target)
+	if err != nil {
+		return abi.NewTokenAmount(0), abi.NewTokenAmount(0), err
+	}
+	if fromVesting.Equals(target) {
+		return fromVesting, abi.NewTokenAmount(0), nil
+	}
+
+	// unlocked funds were just deducted from available, so track that
+	availableBalance = big.Sub(availableBalance, fromVesting)
+	remaining := big.Sub(target, fromVesting)
+
+	fromBalance = big.Min(availableBalance, remaining)
```

Comment:
Call out explicitly that if the penalty exceeds the amount available, the penalty is reduced to match, and point to to issue #627 for bankruptcy.